### DISCUSSION TITLE
Dynamic Components multiple calls

### DISCUSSION THREAD

**Discussion by JSR33** [QuestPDF/QuestPDF#497] [Q&A] [answered]
Reading Dynamic components QuestPDF documentation I found the following sentence: "... The QuestPDF library may perform multiple Compose method calls per page."
Which is the reason for this to happen? I have an extra time consuming due to this and I cannot understand why QuestPDF library needs this.

**Comment by DawidRobertNowak** [comment, marked as answer]
Hello,

We are closing this thread due to inactivity. We apologize for the delayed response. Thank you for taking the time to reach out.

This is expected behavior — QuestPDF uses a multi-pass rendering approach (measuring available space, then drawing content) which requires multiple `Compose` calls per page. To minimize performance impact, we recommend keeping the `Compose` method lightweight and caching any expensive computations outside of it.

You can find the updated documentation here:
https://www.questpdf.com/concepts/code-patterns/dynamic-components.html

Should you encounter any further problems, please don't hesitate to open a new issue.

Thank you!
